### PR TITLE
feat(metrics): add sync device id to access token created events

### PR DESCRIPTION
**Pull request by chenba**
Because:
 - we want to have some idea of when a user's devices were most recently active

This commit:
 - adds the sync device id to the access token created backend Glean event when there's one

### CODE REVIEW

File: packages/fxa-auth-server/lib/metrics/glean/server_events.ts
Comment (chenba) [inline packages/fxa-auth-server/lib/metrics/glean/server_events.ts:1]: File generated by glean_parser.  No need to review.

File: packages/fxa-shared/metrics/glean/fxa-backend-metrics.yaml [hunk 1]
```diff
@@ -150,6 +150,27 @@ session:
...
+    lifetime: ping
+    send_in_pings:
+      - accounts-events
+    notification_emails:
+      - vzare@mozilla.com
+      - fxa-staff@mozilla.com
+    bugs:
+      - https://mozilla-hub.atlassian.net/browse/FXA-9355
+    data_reviews:
+      - https://bugzilla.mozilla.org/show_bug.cgi?id=xxxxxxx
```
Comment (chenba) [inline packages/fxa-shared/metrics/glean/fxa-backend-metrics.yaml:169]: I'll update this URL once I filed a bug for data review.

File: packages/fxa-auth-server/lib/routes/oauth/token.js [hunk 2]
```diff
@@ -380,10 +403,10 @@ module.exports = ({ log, oauthDB, db, mailer, devices, statsd, glean }) => {
       service: oauthClientId,
       uid,
     });
-    glean.oauth.tokenCreated(req, {
+    emitTokenCreated(req, grant, {
```
Comment (vbudhram) [inline packages/fxa-auth-server/lib/routes/oauth/token.js:406]: nit on the name, seems like it should call out its a metric or event

File: packages/fxa-auth-server/test/oauth/routes/token.js [hunk 3]
```diff
@@ -346,10 +369,116 @@ describe('/token POST', function () {
...
+        payload: {
+          client_id: CLIENT_ID,
+          grant_type: 'refresh_token',
+          refresh_token: REFRESH_TOKEN,
+        },
+        emitMetricsEvent: () => {},
+      };
+      await route.config.handler(request);
+      await new Promise((resolve) =>
+        setTimeout(() => {
```
Comment (vbudhram) [inline packages/fxa-auth-server/test/oauth/routes/token.js:406]: Hm I wonder if there is a way to avoid the `setTimeout`

**Review by vbudhram — APPROVED** [review]
@chenba LGTM 👍🏽

File: packages/fxa-auth-server/test/oauth/routes/token.js (hunk 3 above)
Comment (chenba) [reply packages/fxa-auth-server/test/oauth/routes/token.js:406]: I'm not aware of any trick/syntax/api that would make that possible.

**Comment by mhammond** [comment]
At a quick glance it looks like this is the "raw" device ID. All client metrics emit the "hashed" device ID. Is it work making these consistent so we can do more with the ID, such as joining it to the telemetry emitted by clients?

**Comment by chenba** [comment]
> At a quick glance it looks like this is the "raw" device ID. All client metrics emit the "hashed" device ID. Is it work making these consistent so we can do more with the ID, such as joining it to the telemetry emitted by clients?

@mhammond Thank you for taking a look!  We can definitely hash the id.  Is it a SHA-256?

**Comment by mhammond** [comment]
It's hashed with a secret only the servers know, so sadly I don't really know how it's put together in practice. There are some server requests which return it to clients purely for the purposes of these metrics. Maybe @vbudhram has more details?

**Comment by mhammond** [comment]
(It's all a little complicated, because for many years there was a policy to not ever use the raw device ID. In the last 12 months however that policy has changed, so depending on how it's being used it *might* be OK. So it's not clear that using this hashed ID is the right thing to do, but it *is* the right thing to do it cross-referencing telemetry collected by clients is necessary. I wish I could offer better clarity!)

**Comment by chenba** [comment]
After poking around a bit, it looks like 
 - Firefox Desktop is not using Glean yet, and the device id in the metrics is a hash with the raw device id and the `identity.fxaccounts.account.telemetry.sanitized_uid` pref value.  The sanitized uid is the value that's been HMAC-ed with a secret.
 - Mobile Firefox appears to be using Glean, and the UUID it uses, https://dictionary.telemetry.mozilla.org/apps/fenix/metrics/sync_v2_sync_uuid, is simply a UUID generated with the glean.js SDK.

**Comment by mhammond** [comment]
> * Mobile Firefox appears to be using Glean, and the UUID it uses, https://dictionary.telemetry.mozilla.org/apps/fenix/metrics/sync_v2_sync_uuid, is simply a UUID generated with the glean.js SDK.

[here](https://github.com/mozilla/application-services/blob/a69ebea12384421435ff6b555573e04dc497dcdc/components/sync_manager/metrics.yaml#L66-L69) is the hashed uid in glean for mobile.

**Comment by chenba** [comment]
I think a setup that correlate the various identifiers among multiple datasets is beyond the scope of this PR.

> > * Mobile Firefox appears to be using Glean, and the UUID it uses, https://dictionary.telemetry.mozilla.org/apps/fenix/metrics/sync_v2_sync_uuid, is simply a UUID generated with the glean.js SDK.
> 
> [here](https://github.com/mozilla/application-services/blob/a69ebea12384421435ff6b555573e04dc497dcdc/components/sync_manager/metrics.yaml#L66-L69) is the hashed device id in glean for mobile.

I think that's just the sanitized uid?

**Comment by mhammond** [comment]
I hope not - our sync code [goes to effort to capture it from the token server](https://github.com/mozilla/application-services/blob/a69ebea12384421435ff6b555573e04dc497dcdc/components/sync15/src/client/token.rs#L382C8-L384) and [sticks it in the "telemetry" data sent back to the mobile apps](https://github.com/mozilla/application-services/blob/a69ebea12384421435ff6b555573e04dc497dcdc/components/sync15/src/telemetry.rs#L748) - it's been a while since I've verified that all still makes it through correctly but we'd consider it a bug otherwise.

**Comment by chenba** [comment]
hmm I understand that the hashed ("sanitized") uid needs to be fetched from the tokenserver, but I thought we were talking about [this device id](https://searchfox.org/mozilla-central/source/services/fxaccounts/FxAccountsTelemetry.sys.mjs#96), which was what I couldn't find in the mobile Glean pings.

**Comment by mhammond** [comment]
There's a little more detail in [this bugzilla comment](https://bugzilla.mozilla.org/show_bug.cgi?id=1604844#c3) - it's all ripe for a bit of a refresh :)

File: packages/fxa-auth-server/test/oauth/routes/token.js [hunk 4]
```diff
@@ -346,10 +369,116 @@ describe('/token POST', function () {
...
+          sinon.assert.calledOnceWithExactly(
+            mockGlean.oauth.tokenCreated,
+            request,
+            {
+              uid: UID,
+              oauthClientId: CLIENT_ID,
+              reason: 'authorization_code',
+              syncDeviceId: '42cbcdce5c0a84a532b2722633af9f03',
```
Comment (pdehaan) [inline packages/fxa-auth-server/test/oauth/routes/token.js:381]: Should this be `mockDevice.findDeviceIdByUidAndSessionId`?

https://github.com/mozilla/fxa/blob/d8441c6cfe6a5bfd7fd7ede6dac3cf483e9f634e/packages/fxa-auth-server/test/oauth/routes/token.js#L28-L29

File: packages/fxa-auth-server/test/oauth/routes/token.js [hunk 5]
```diff
@@ -346,10 +369,116 @@ describe('/token POST', function () {
...
+      await new Promise((resolve) =>
+        setTimeout(() => {
+          sinon.assert.calledOnceWithExactly(
+            mockGlean.oauth.tokenCreated,
+            request,
+            {
+              uid: UID,
+              oauthClientId: CLIENT_ID,
+              reason: 'refresh_token',
+              syncDeviceId: 'e553a430d7e5b904b1edd15a33af9f03',
```
Comment (pdehaan) [inline packages/fxa-auth-server/test/oauth/routes/token.js:414]: Should this be `mockDevice.findDeviceIdByUidAndRefreshTokenId`:
https://github.com/mozilla/fxa/blob/d8441c6cfe6a5bfd7fd7ede6dac3cf483e9f634e/packages/fxa-auth-server/test/oauth/routes/token.js#L28-L32

File: packages/fxa-auth-server/test/oauth/routes/token.js [hunk 6]
```diff
@@ -346,10 +369,116 @@ describe('/token POST', function () {
...
+      await new Promise((resolve) =>
+        setTimeout(() => {
+          sinon.assert.calledOnceWithExactly(
+            mockGlean.oauth.tokenCreated,
+            request,
+            {
+              uid: UID,
+              oauthClientId: CLIENT_ID,
+              reason: 'fxa-credentials',
+              syncDeviceId: '42cbcdce5c0a84a532b2722633af9f03',
```
Comment (pdehaan) [inline packages/fxa-auth-server/test/oauth/routes/token.js:441]: `mockDevice.findDeviceIdByUidAndSessionId`?

File: packages/fxa-auth-server/test/oauth/routes/token.js (hunk 4 above)
Comment (chenba) [reply packages/fxa-auth-server/test/oauth/routes/token.js:381]: No, the return value is used not the function itself.

**Comment by chenba** [comment]
Closing this as we are not likely to move forward with including plain device ids in the glean pings.

**Comment by mhammond** [comment]
FWIW, the client code is trying to move towards recording these raw IDs in glean pings - https://bugzilla.mozilla.org/show_bug.cgi?id=1893727 is for the uid, but IIUC we ultimately intend the same for device IDs. cc @tarikeshaq

**Pull request closed without merging**